conn.go

```diff
@@ -0,0 +1,214 @@
+package coap
+
+import (
+	"net"
+	"sync/atomic"
+	"time"
+)
+
+type writeReq interface {
+}
+
+type writeReqBase struct {
+	req      Message
+	respChan chan error
+}
+
+type writeReqTCP struct {
+	writeReqBase
+}
+
+type writeReqUDP struct {
+	writeReqBase
+	sessionData *SessionUDPData
+}
+
+type conn interface {
+	LocalAddr() net.Addr
+	RemoteAddr() net.Addr
+	Write(w writeReq) error
+	Close() error
+}
+
+type connBase struct {
+	writeChan chan writeReq
+	closeChan chan bool
+	finChan   chan bool
+	closed    int32
+}
+
+func (conn *connBase) finishWrite() {
+	if !atomic.CompareAndSwapInt32(&conn.closed, conn.closed, 1) {
+		return
+	}
+	conn.closeChan <- true
+	<-conn.finChan
+}
+
+type connTCP struct {
+	connBase
+	connection net.Conn // i/o connection if TCP was used
+	num        int32
+}
+
+func (conn *connTCP) LocalAddr() net.Addr {
+	return conn.connection.LocalAddr()
+}
+
+func (conn *connTCP) RemoteAddr() net.Addr {
+	return conn.connection.RemoteAddr()
+}
+
+func (conn *connTCP) Close() error {
+	conn.finishWrite()
+	return conn.connection.Close()
+}
+
+func (conn *connTCP) Write(w writeReq) error {
+	if atomic.LoadInt32(&conn.closed) > 0 {
+		return ErrConnectionClosed
+	}
+	conn.writeChan <- w
+	err := <-w.(*writeReqTCP).respChan
+	return err
+}
+
+type connUDP struct {
+	connBase
+	connection *net.UDPConn // i/o connection if UDP was used
+}
+
+func (conn *connUDP) LocalAddr() net.Addr {
+	return conn.connection.LocalAddr()
+}
+
+func (conn *connUDP) RemoteAddr() net.Addr {
+	return conn.connection.RemoteAddr()
+}
+
+func (conn *connUDP) Write(w writeReq) error {
+	if atomic.LoadInt32(&conn.closed) > 0 {
+		return ErrConnectionClosed
+	}
+	conn.writeChan <- w
+	return <-w.(*writeReqUDP).respChan
+}
+
+func (conn *connUDP) SetReadDeadline(timeout time.Time) error {
+	return conn.connection.SetReadDeadline(timeout)
+}
+
+func (conn *connUDP) ReadFromSessionUDP(m []byte) (int, *SessionUDPData, error) {
+	return ReadFromSessionUDP(conn.connection, m)
+}
+
+func (conn *connUDP) Close() error {
+	conn.finishWrite()
+	return conn.connection.Close()
+}
+
+func newConnectionTCP(c net.Conn, srv *Server) conn {
+	connection := &connTCP{connBase: connBase{writeChan: make(chan writeReq), closeChan: make(chan bool), finChan: make(chan bool), closed: 0}, connection: c}
+	go writeTCP(connection, srv)
+	return connection
+}
+
+func newConnectionUDP(c *net.UDPConn, srv *Server) conn {
+	connection := &connUDP{connBase: connBase{writeChan: make(chan writeReq), closeChan: make(chan bool), finChan: make(chan bool), closed: 0}, connection: c}
+	go writeUDP(connection, srv)
+	return connection
+}
+
+func writeTCP(conn *connTCP, srv *Server) {
+	wr := srv.acquireWriter(conn.connection)
+LOOP:
```

Comment:
try to avoid these, while this is perfectly fine in C and valid in Go, common practice is to extract functionality into multiple funcs and return instead of break ANCH.

Reply:
extracted to function: FIXED